File: lib/prime-project.py
"""Insert or replace FLOW priming content in a project's CLAUDE.md.

Reads frameworks/<name>/priming.md and inserts it between
<!-- FLOW:BEGIN --> / <!-- FLOW:END --> markers in the target
project's CLAUDE.md. Idempotent — re-running replaces existing
primed content.

Usage: bin/flow prime-project <project_root> --framework <name>

Output (JSON to stdout):
  {"status": "ok", "framework": "...", "project_root": "..."}
"""

import json
import sys
from pathlib import Path


MARKER_BEGIN = "<!-- FLOW:BEGIN -->"
MARKER_END = "<!-- FLOW:END -->"


def _frameworks_dir():
    """Return the frameworks/ directory relative to this script."""
    return Path(__file__).resolve().parent.parent / "frameworks"


def prime(project_root, framework, frameworks_dir=None):
    """Insert or replace priming content in CLAUDE.md."""
    if frameworks_dir is None:
        frameworks_dir = str(_frameworks_dir())

    project = Path(project_root)
    claude_md = project / "CLAUDE.md"

    if not claude_md.exists():
        return {
            "status": "error",
            "message": "CLAUDE.md not found in project root",
        }

    priming_path = Path(frameworks_dir) / framework / "priming.md"
    if not priming_path.exists():
        return {
            "status": "error",
            "message": f"Framework not found: {framework}",
        }

    priming_content = priming_path.read_text()
    existing_content = claude_md.read_text()

    block = f"{MARKER_BEGIN}\n{priming_content}\n{MARKER_END}\n"

    if MARKER_BEGIN in existing_content and MARKER_END in existing_content:
        begin_index = existing_content.index(MARKER_BEGIN)
        end_index = existing_content.index(MARKER_END) + len(MARKER_END)
        if end_index < len(existing_content) and existing_content[end_index] == "\n":
            end_index += 1
        new_content = existing_content[:begin_index] + block + existing_content[end_index:]
    else:
        new_content = existing_content + "\n" + block

    claude_md.write_text(new_content)

    return {
        "status": "ok",
        "framework": framework,
        "project_root": str(project_root),
    }


def main():
    if len(sys.argv) < 2:
        print(json.dumps({
            "status": "error",
            "message": "Usage: bin/flow prime-project <project_root> --framework <name>",
        }))
        sys.exit(1)

    project_root = sys.argv[1]
    framework = None

    for i, arg in enumerate(sys.argv[2:], start=2):
        if arg == "--framework" and i + 1 < len(sys.argv):
            framework = sys.argv[i + 1]
            break

    if not framework:
        print(json.dumps({
            "status": "error",
            "message": "Missing --framework argument",
        }))
        sys.exit(1)

    result = prime(project_root, framework)
    print(json.dumps(result))
    if result["status"] == "error":
        sys.exit(1)


if __name__ == "__main__":
    main()

File: src/prime_project.rs
//! Insert or replace FLOW priming content in a project's CLAUDE.md.
//!
//! Reads
//! `frameworks/<name>/priming.md` and inserts it between
//! `<!-- FLOW:BEGIN -->` / `<!-- FLOW:END -->` markers in the target
//! project's CLAUDE.md. Idempotent — re-running replaces existing
//! primed content.
//!
//! Usage: `bin/flow prime-project <project_root> --framework <name>`
//!
//! Output (JSON to stdout):
//!   `{"status": "ok", "framework": "...", "project_root": "..."}`
//!   `{"status": "error", "message": "..."}`

use std::fs;
use std::path::{Path, PathBuf};
use std::process;

use clap::Args as ClapArgs;
use serde_json::{json, Value};

use crate::utils::frameworks_dir;

const MARKER_BEGIN: &str = "<!-- FLOW:BEGIN -->";
const MARKER_END: &str = "<!-- FLOW:END -->";

#[derive(ClapArgs)]
pub struct Args {
    /// Project root directory (contains CLAUDE.md)
    pub project_root: String,

    /// Framework name (rails, python, ios, go, rust)
    #[arg(long)]
    pub framework: String,
}

/// Insert or replace priming content in a project's CLAUDE.md.
///
/// Returns a JSON Value with the same `status`, `replaced`, and
/// `message` fields the prime skill expects when reporting the
/// outcome to the user.
///
/// # Marker handling
///
/// Marker detection uses byte offsets. The markers
/// (`<!-- FLOW:BEGIN -->` / `<!-- FLOW:END -->`) are ASCII only, so
/// byte-index slicing never splits a multi-byte UTF-8 character. This
/// invariant is asserted via a debug_assert for defense in depth.
pub fn prime(project_root: &Path, framework: &str, fw_dir: &Path) -> Value {
    let claude_md = project_root.join("CLAUDE.md");
    if !claude_md.exists() {
        return json!({
            "status": "error",
            "message": "CLAUDE.md not found in project root",
        });
    }

    let priming_path = fw_dir.join(framework).join("priming.md");
    if !priming_path.exists() {
        return json!({
            "status": "error",
            "message": format!("Framework not found: {}", framework),
        });
    }

    let priming_content = match fs::read_to_string(&priming_path) {
        Ok(c) => c,
        Err(e) => {
            return json!({
                "status": "error",
                "message": format!("Could not read priming.md: {}", e),
            })
        }
    };

    let existing_content = match fs::read_to_string(&claude_md) {
        Ok(c) => c,
        Err(e) => {
            return json!({
                "status": "error",
                "message": format!("Could not read CLAUDE.md: {}", e),
            })
        }
    };

    // ASCII-only markers make byte-index slicing safe.
    debug_assert!(MARKER_BEGIN.is_ascii() && MARKER_END.is_ascii());

    let block = format!("{}\n\n{}\n{}\n", MARKER_BEGIN, priming_content, MARKER_END);

    let new_content = match (
        existing_content.find(MARKER_BEGIN),
        existing_content.find(MARKER_END),
    ) {
        (Some(begin_index), Some(end_index)) => {
            let mut end = end_index + MARKER_END.len();
            if existing_content.as_bytes().get(end) == Some(&b'\n') {
                end += 1;
            }
            let mut out = String::with_capacity(existing_content.len() + block.len());
            out.push_str(&existing_content[..begin_index]);
            out.push_str(&block);
            out.push_str(&existing_content[end..]);
            out
        }
        _ => {
            let mut out = String::with_capacity(existing_content.len() + block.len() + 1);
            out.push_str(&existing_content);
            out.push('\n');
            out.push_str(&block);
            out
        }
    };

    if let Err(e) = fs::write(&claude_md, new_content) {
        return json!({
            "status": "error",
            "message": format!("Could not write CLAUDE.md: {}", e),
        });
    }

    json!({
        "status": "ok",
        "framework": framework,
        "project_root": project_root.display().to_string(),
    })
}

/// Build the CLI result as a JSON value. Returns `Err` when the result
/// `status` is `"error"` so `run` can exit non-zero while still
/// printing the JSON body.
pub fn run_impl(args: &Args) -> Result<Value, Value> {
    let project_root = PathBuf::from(&args.project_root);
    let fw_dir = match frameworks_dir() {
        Some(p) => p,
        None => {
            return Err(json!({
                "status": "error",
                "message": "Plugin root not found",
            }))
        }
    };
    let result = prime(&project_root, &args.framework, &fw_dir);
    if result.get("status").and_then(|v| v.as_str()) == Some("error") {
        Err(result)
    } else {
        Ok(result)
    }
}

pub fn run(args: Args) {
    match run_impl(&args) {
        Ok(value) => {
            println!("{}", serde_json::to_string(&value).unwrap());
        }
        Err(value) => {
            println!("{}", serde_json::to_string(&value).unwrap());
            process::exit(1);
        }
    }
}
